### PR TITLE
implement Sum for Amount

### PR DESCRIPTION
**Pull request by oxarbitrage**
## Motivation

All transaction versions can include transparent inputs and outputs. The transparent value balance for the transaction is "the sum of the outputs spent by transparent inputs in `tx_in` fields, minus the sum of newly created outputs in `tx_out` fields (https://github.com/ZcashFoundation/zebra/blob/main/book/src/dev/rfcs/0012-value-pools.md#definitions).

### Designs

This a tiny part of https://github.com/ZcashFoundation/zebra/blob/main/book/src/dev/rfcs/0012-value-pools.md

## Solution

Create a `Sum` operator for `Amount` that will help us calculating the transparent value balance.

## Review

Anyone can review, is a simple addition but @teor2345 may have more context than others.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

## Follow Up Work

A next PR will implement value balance type.

### CODE REVIEW

File: zebra-chain/src/amount.rs [hunk 1]
```diff
@@ -293,6 +293,15 @@ impl std::ops::Div<u64> for Amount<NonNegative> {
     }
 }
 
+impl<C> std::iter::Sum<Amount<C>> for Result<Amount<C>>
+where
+    C: Constraint,
+{
+    fn sum<I: Iterator<Item = Amount<C>>>(iter: I) -> Self {
+        Amount::try_from(iter.map(|a| a.0).fold(0i64, std::ops::Add::add))
```
Comment (teor2345) [inline zebra-chain/src/amount.rs]: What happens if this sum overflows `i64::MAX`? Or underflows `i64::MIN`?

Can you add a test for this kind of overflow?

File: zebra-chain/src/amount.rs [hunk 2]
```diff
@@ -701,4 +710,56 @@ mod test {
...
+        zebra_test::init();
+
+        let one: Amount = 1.try_into()?;
+        let neg_one: Amount = (-1).try_into()?;
+
+        let zero: Amount = 0.try_into()?;
+
+        // success
+        let amounts = vec![one, neg_one, zero];
+        let sum: Amount = amounts.iter().copied().sum::<Result<Amount, Error>>()?;
```
Comment (teor2345) [inline zebra-chain/src/amount.rs]: Do you also want to `impl<C> std::iter::Sum<&Amount<C>> for Result<Amount<C>>`?
That way, you don't have to add `.copied()` all the time.

This change is totally optional.

File: zebra-chain/src/amount.rs [hunk 3]
```diff
@@ -701,4 +710,56 @@ mod test {
 
         Ok(())
     }
+
+    #[test]
+    fn test_sum() -> Result<()> {
```
Comment (teor2345) [inline zebra-chain/src/amount.rs:742]: These are good tests, thanks!

**Review by teor2345 — CHANGES_REQUESTED** [review]
Looks like a good initial implementation and tests.

What happens if the sum overflows `i64`?

I also made an optional suggestion that might make writing other code easier.

File: zebra-chain/src/amount.rs (hunk 2 above)
Comment (oxarbitrage) [reply zebra-chain/src/amount.rs]: I think this is a good idea, i actually tried to do it but had compiler problems with the reference lifetimes that i was not able to resolve. Will need some assistance there.

File: zebra-chain/src/amount.rs [hunk 4]
```diff
@@ -293,6 +301,29 @@ impl std::ops::Div<u64> for Amount<NonNegative> {
...
+        let mut overflow = false;
+        let sum = iter
+            .map(|a| a.0)
+            .fold(0i64, |acc, amount| match acc.checked_add(amount) {
+                Some(result) => result,
+                None => {
+                    overflow = true;
+                    0
+                }
+            });
```
Comment (teor2345) [inline zebra-chain/src/amount.rs]: Suggestion: Do you want to simplify this code using `try_fold`?
https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.try_fold

This change is totally optional, but it might make the code easier to understand.

File: zebra-chain/src/amount.rs [hunk 5]
```diff
@@ -311,6 +342,9 @@ pub enum Error {
     MultiplicationOverflow { amount: i64, multiplier: u64 },
     /// cannot divide amount {amount} by zero
     DivideByZero { amount: i64 },
+
+    /// Attempt to sum with overflow
```
Comment (teor2345) [inline zebra-chain/src/amount.rs]: Can we try to match the wording of the other overflow error?
(These comments are used as error messages by the `thiserror` derive.)

```suggestion
    /// i64 overflow when summing i64 amounts
```

File: zebra-chain/src/amount.rs [hunk 6]
```diff
@@ -47,6 +47,14 @@ impl<C> Amount<C> {
         LittleEndian::write_i64(&mut buf, self.0);
         buf
     }
+
+    /// Create a zero `Amount`
+    pub fn zero() -> Amount<C>
```
Comment (teor2345) [inline zebra-chain/src/amount.rs:52]: This is a good change, thanks!

File: zebra-chain/src/amount.rs [hunk 7]
```diff
@@ -701,4 +735,80 @@ mod test {
...
+            Err(e) => e,
+            _ => unreachable!("above operation will always fail"),
+        };
+        assert_eq!(
+            err,
+            Error::Contains {
+                range: -MAX_MONEY..=MAX_MONEY,
+                value: integer_sum
+            }
+        );
```
Comment (teor2345) [inline zebra-chain/src/amount.rs]: Suggestion: `assert_eq!` prints both values, so we will get better error messages if we compare the result with `Err(Error::Contains { ... })`.

Can you also make this change for the other error asserts?

File: zebra-chain/src/amount.rs [hunk 8]
```diff
@@ -701,4 +735,80 @@ mod test {
...
+
+        // above max of i64 error
+        let times = i64::MAX / MAX_MONEY;
+        let mut amounts = vec![MAX_MONEY.try_into()?];
+        for _ in 0..times {
+            amounts.push(MAX_MONEY.try_into()?);
+        }
```
Comment (teor2345) [inline zebra-chain/src/amount.rs]: Suggestion: Do you want to simplify this code using `iter::repeat`?
https://doc.rust-lang.org/std/iter/fn.repeat.html

This change is totally optional, but it should make the code easier to read.

**Review by teor2345 — CHANGES_REQUESTED** [review]
I made a few suggestions that are optional.

I'll go look into the lifetimes now, and see what I can come up with.

You can also try using the latest stable Rust compiler, they are always adding more hints for lifetimes and other tricky code.

File: zebra-chain/src/amount.rs (hunk 2 above)
Comment (teor2345) [reply zebra-chain/src/amount.rs]: I opened PR #2506, you're right, it was a bit tricky.

File: zebra-chain/src/amount.rs (hunk 4 above)
Comment (oxarbitrage) [reply zebra-chain/src/amount.rs]: https://github.com/ZcashFoundation/zebra/pull/2500/commits/50f41e874f4cc8ae7244d7b252d608891943d4b2

File: zebra-chain/src/amount.rs (hunk 5 above)
Comment (oxarbitrage) [reply zebra-chain/src/amount.rs]: https://github.com/ZcashFoundation/zebra/pull/2500/commits/81b4e43235fa192183896ec7e0d90cff27760929

File: zebra-chain/src/amount.rs (hunk 6 above)
Comment (oxarbitrage) [reply zebra-chain/src/amount.rs:52]: Yes, is handy to have.

File: zebra-chain/src/amount.rs (hunk 7 above)
Comment (oxarbitrage) [reply zebra-chain/src/amount.rs]: https://github.com/ZcashFoundation/zebra/pull/2500/commits/a0149389344d2d304a239af27847af243c3c1944

File: zebra-chain/src/amount.rs (hunk 8 above)
Comment (oxarbitrage) [reply zebra-chain/src/amount.rs]: https://github.com/ZcashFoundation/zebra/pull/2500/commits/d71082b474adf85110ff5b8a0b1dde08ee4c37a6

File: zebra-chain/src/amount.rs (hunk 2 above)
Comment (oxarbitrage) [reply zebra-chain/src/amount.rs]: Thanks, that was not an easy job. I merged your branch into this one.

**Review by teor2345 — APPROVED** [review]
Thanks, all those changes look great!
